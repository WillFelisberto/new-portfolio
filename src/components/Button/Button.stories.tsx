import { Meta, StoryObj } from '@storybook/react';
import Button from '.';

const meta: Meta<typeof Button> = {
  title: 'Components/Button',
  component: Button,
  argTypes: {
    children: {
      control: 'text',
      description: 'Texto ou conteúdo dentro do botão.',
      defaultValue: 'Click me!',
    },
    onClick: { action: 'clicked' },
    className: {
      control: 'text',
      description: 'Classes CSS adicionais para estilização personalizada.',
    },
    disabled: {
      control: 'boolean',
      description: 'Desativa o botão.',
    },
  },
} satisfies Meta<typeof Button>;

export default meta;
type Story = StoryObj<typeof Button>;

export const Default: Story = {
  args: {
    children: 'Click me!',
    className: '',
    disabled: false,
  },
};

export const WithCustomClass: Story = {
  args: {
    children: 'Custom Class Button',
    className: 'bg-red-500 hover:bg-red-400',
  },
};

export const Disabled: Story = {
  args: {
    children: 'Disabled Button',
    disabled: true,
  },
};

export const WithIcon: Story = {
  args: {
    children: (
      <>
        <span>Icon Button</span>
        <svg
          xmlns="http://www.w3.org/2000/svg"
          fill="none"
          viewBox="0 0 24 24"
          strokeWidth={1.5}
          stroke="currentColor"
          className="h-4 w-4"
        >
          <path
            strokeLinecap="round"
            strokeLinejoin="round"
            d="M15.75 9V5.25a2.25 2.25 0 10-4.5 0V9m-6 0h15M4.5 9h15M4.5 15.75h15m-15 0h15M4.5 20.25h15m-15 0h15M4.5 15.75h15m-15 0H7.5"
          />
        </svg>
      </>
    ),
  },
};
