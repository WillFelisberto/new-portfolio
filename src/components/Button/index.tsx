import { cn } from '@/lib/utils';
import { ButtonHTMLAttributes } from 'react';

type ButtonProps = ButtonHTMLAttributes<HTMLButtonElement>;

const Button = ({ children, className, ...props }: ButtonProps) => {
  return (
    <button
      className={cn(
        'flex items-center justify-center gap-2 rounded-lg bg-sky-600 px-4 py-3 text-gray-50 transition-all hover:bg-sky-500 disabled:opacity-50',
        className,
      )}
      {...props}
    >
      {children}
    </button>
  );
};

export default Button;
