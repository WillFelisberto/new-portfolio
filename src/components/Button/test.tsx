import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Button from '.';

describe('<Button />', () => {
  it('should render the button with the correct text', () => {
    const { container } = render(<Button>Click me!</Button>);
    const button = screen.getByRole('button', { name: /click me!/i });
    expect(button).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should apply default styles', () => {
    render(<Button>Click me!</Button>);
    const button = screen.getByRole('button', { name: /click me!/i });
    expect(button).toHaveClass(
      'flex items-center justify-center gap-2 rounded-lg bg-sky-600 px-4 py-3 text-gray-50 transition-all hover:bg-sky-500 disabled:opacity-50',
    );
  });

  it('should accept and apply custom class names', () => {
    render(<Button className="custom-class">Click me!</Button>);
    const button = screen.getByRole('button', { name: /click me!/i });
    expect(button).toHaveClass('custom-class');
  });

  it('should be clickable and call the onClick handler', async () => {
    const handleClick = jest.fn();
    render(<Button onClick={handleClick}>Click me!</Button>);
    const button = screen.getByRole('button', { name: /click me!/i });

    await userEvent.click(button);
    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  it('should render as disabled when disabled prop is passed', () => {
    render(<Button disabled>Click me!</Button>);
    const button = screen.getByRole('button', { name: /click me!/i });
    expect(button).toBeDisabled();
    expect(button).toHaveClass('disabled:opacity-50');
  });

  it('should not call onClick when disabled', async () => {
    const handleClick = jest.fn();
    render(
      <Button onClick={handleClick} disabled>
        Click me!
      </Button>,
    );
    const button = screen.getByRole('button', { name: /click me!/i });

    await userEvent.click(button);
    expect(handleClick).not.toHaveBeenCalled();
  });

  it('should render children correctly', () => {
    render(
      <Button>
        <span>Child Text</span>
      </Button>,
    );
    const childText = screen.getByText(/child text/i);
    expect(childText).toBeInTheDocument();
  });
});
