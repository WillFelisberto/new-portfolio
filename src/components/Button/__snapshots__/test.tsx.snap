// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render the button with the correct text 1`] = `
<button
  class="flex items-center justify-center gap-2 rounded-lg bg-sky-600 px-4 py-3 text-gray-50 transition-all hover:bg-sky-500 disabled:opacity-50"
>
  Click me!
</button>
`;
